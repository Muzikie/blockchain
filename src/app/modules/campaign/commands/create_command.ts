/* eslint-disable class-methods-use-this */
import { Modules, StateMachine } from 'klayr-sdk';
import { CampaignStore } from '../stores/campaign';
import { CampaignAccountStore } from '../stores/campaign_account';
import { CampaignCreated } from '../events/campaign_created';
import { createCommandParamsSchema } from '../schemas';
import { CreateCommandParams, Campaign, CampaignStatus, CampaignAccount } from '../types';
import { getCampaignId } from '../utils';

export class CreateCommand extends Modules.BaseCommand {
	public async verify(
		context: StateMachine.CommandVerifyContext<CreateCommandParams>,
	): Promise<StateMachine.VerificationResult> {
		const { params } = context;
		const campaignStore = this.stores.get(CampaignStore);
		const campaignId = getCampaignId(params);

		const campaignExists = await campaignStore.has(context, campaignId);
		if (campaignExists) {
			throw new Error('This campaign already exist.');
		}

		return { status: StateMachine.VerifyStatus.OK };
	}

	public async execute(
		context: StateMachine.CommandExecuteContext<CreateCommandParams>,
	): Promise<void> {
		const {
			params,
			transaction: { senderAddress },
		} = context;

		const campaignAccountStore = this.stores.get(CampaignAccountStore);
		const campaignStore = this.stores.get(CampaignStore);

		// Create campaign ID
		const campaignId = getCampaignId(params);

		// Create campaign object
		const campaign: Campaign = {
			softGoal: BigInt(params.softGoal),
			hardGoal: BigInt(params.hardGoal),
			currentFunding: BigInt(0),
			status: CampaignStatus.Draft,
			deadline: params.deadline,
			apiId: params.apiId,
			contributionTiers: [],
			id: campaignId,
			submitter: senderAddress,
		};

		// Store the campaign object in the blockchain
		await campaignStore.set(context, campaignId, campaign);

		// Get the sender account from the blockchain
		const senderExists = await campaignAccountStore.has(context, senderAddress);
		let senderAccount: CampaignAccount;
		if (!senderExists) {
			senderAccount = {
				campaigns: [campaignId],
			};
		} else {
			const retrievedAccount = await campaignAccountStore.get(context, senderAddress);
			senderAccount = {
				campaigns: [...retrievedAccount.campaigns, campaignId],
			};
		}

		// Store the account object in the blockchain
		await campaignAccountStore.set(context, senderAddress, senderAccount);

		const campaignCreated = this.events.get(CampaignCreated);
		campaignCreated.add(
			context,
			{
				submitter: senderAddress,
				campaignId,
			},
			[senderAddress],
		);
	}

	public schema = createCommandParamsSchema;
}
