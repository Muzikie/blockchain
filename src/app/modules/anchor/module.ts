/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  TokenMethod,
  // TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { VerifyStatus } from 'lisk-framework';
import {
  accountStoreSchema,
  addressRequestSchema,
  idRequestSchema,
  anchorStoreSchema,
  hasAnchorResponse,
} from './schemas';
import { AnchorEndpoint } from './endpoint';
import { AnchorMethod } from './method';
import { AnchorStore } from './stores/anchor';
import { AnchorAccountStore } from './stores/anchorAccount';
import { AnchorCreated } from './events/anchorCreated';
import { CreateCommand } from './commands/create_command';
import { VoteCommand } from './commands/vote_command';

export class AnchorModule extends BaseModule {
  public endpoint = new AnchorEndpoint(this.stores, this.offchainStores);
  public method = new AnchorMethod(this.stores, this.events);

  private readonly _createCommand = new CreateCommand(this.stores, this.events);
  private readonly _purchaseCommand = new VoteCommand(this.stores, this.events);

  // eslint-disable-next-line @typescript-eslint/member-ordering
  public commands = [this._createCommand, this._purchaseCommand];

  private _tokenMethod!: TokenMethod;

  public constructor() {
    super();
    this.stores.register(AnchorAccountStore, new AnchorAccountStore(this.name, 0));
    this.stores.register(AnchorStore, new AnchorStore(this.name, 1));
    this.events.register(AnchorCreated, new AnchorCreated(this.name));
  }

  public addDependencies(tokenMethod: TokenMethod): void {
    this._tokenMethod = tokenMethod;

    this._purchaseCommand.addDependencies(this._tokenMethod);
g  }

  public metadata(): ModuleMetadata {
    return {
      ...this.baseMetadata(),
      endpoints: [
        {
          name: this.endpoint.getAccount.name,
          request: addressRequestSchema,
          response: accountStoreSchema,
        },
        {
          name: this.endpoint.getAnchor.name,
          request: idRequestSchema,
          response: anchorStoreSchema,
        },
        {
          name: this.endpoint.hasAnchor.name,
          request: addressRequestSchema,
          response: hasAnchorResponse,
        },
      ],
      assets: [],
    };
  }

  // Lifecycle hooks
  public async init(_args: ModuleInitArgs): Promise<void> {
    // initialize this module when starting a node
  }

  public async insertAssets(_context: InsertAssetContext) {
    // initialize block generation, add asset
  }

  public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
    // verify block
  }

  // Lifecycle hooks
  // eslint-disable-next-line @typescript-eslint/require-await
  public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
    // verify transaction will be called multiple times in the transaction pool
    return {
      status: VerifyStatus.OK,
    };
  }

  // public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {}

  // public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
