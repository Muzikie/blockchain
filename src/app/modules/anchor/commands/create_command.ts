/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AnchorStore } from '../stores/anchor';
import { AnchorAccountStore } from '../stores/anchorAccount';
import { AnchorCreated } from '../events/anchorCreated';
import { CreateCommandParams, Anchor, AnchorAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { getCreatedAt, getAnchorID } from '../utils';
import { VOTE_RATE_LIMIT } from '../constants';
import { BadgeMethod } from '../../badge/method';
import { Badges } from '../../badge/types';
import { getBadgeID } from '../../badge/utils';
import { AnchorStatsStore } from '../stores/anchorStats';

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;
  private _badgeMethod!: BadgeMethod;

  public addDependencies(badgeMethod: BadgeMethod) {
    this._badgeMethod = badgeMethod;
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const anchorStore = this.stores.get(AnchorStore);
    const { senderAddress } = context.transaction;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorId = getAnchorID(context.params);

    const anchorExists = await anchorStore.has(context, anchorId);
    if (anchorExists) {
      throw new Error('This anchor already exist.');
    }

    // Add submission rate limit
    const senderExists = await anchorAccountStore.has(context, senderAddress);
    if(senderExists) {
      const senderAccount = await anchorAccountStore.get(context, senderAddress);
      const IDS = senderAccount.anchors.slice(-1 * VOTE_RATE_LIMIT);

      if(IDS.length >= VOTE_RATE_LIMIT) {
        const anchor = await anchorStore.get(context, IDS[0]);

        if (anchor.createdAt === getCreatedAt(Math.floor(((new Date()).getTime())))) {
          throw new Error(`You have exceeded the ${VOTE_RATE_LIMIT} anchor submissions daily limit.`);
        }
      }
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const {
      params,
      transaction: { senderAddress },
    } = context;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorStore = this.stores.get(AnchorStore);
    const anchorStatsStore = this.stores.get(AnchorStatsStore);
    const methodContext = context.getMethodContext();

    // Create anchor ID
    const anchorID = getAnchorID(context.params);
    const createdAt = getCreatedAt(new Date().getTime());
    // Create anchor object
    const anchor: Anchor = {
      ...params,
      id: anchorID,
      createdAt,
      votes: [],
      submitter: senderAddress,
    };

    const badgeIDs = [1, 2, 3]
      .map((rank) => getBadgeID(createdAt, rank, Badges.AOTD));
    context.logger.info(badgeIDs.map((idd) => idd.toString('hex')));
    // Store the anchor object in the blockchain
    await anchorStore.set(context, anchorID, anchor);

    const StatsExist = await anchorStatsStore.has(context, Buffer.from(createdAt));
  
    if (!StatsExist) {
      const newAnchorStats = {
        date: createdAt,
        anchorsCount: 1,
        votesCount: 0,
      };
      await anchorStatsStore.set(context, Buffer.from(createdAt), newAnchorStats);
      context.logger.info(`Anchors count for ${createdAt}: 1`);
    } else {
      const anchorStats = await anchorStatsStore.get(context, Buffer.from(createdAt));
      anchorStats.anchorsCount += 1;
      await anchorStatsStore.set(context, Buffer.from(createdAt), anchorStats);
      context.logger.info(`Anchors count2SS for ${createdAt}: ${anchorStats.anchorsCount}`);
    }

    // Get the sender account from the blockchain
    const senderExists = await anchorAccountStore.has(context, senderAddress);
    let senderAccount: AnchorAccount;
    if (!senderExists) {
      senderAccount = {
        anchors: [anchorID],
        votes: [],
      };
    } else {
      const retrievedAccount = await anchorAccountStore.get(context, senderAddress);
      senderAccount = {
        anchors: [...retrievedAccount.anchors, anchorID],
        votes: retrievedAccount.votes
      };
    }

    // Store the account object in the blockchain
    await anchorAccountStore.set(context, senderAddress, senderAccount);

    await this._badgeMethod.createBadgesForDay(methodContext, createdAt);

    const anchorCreated = this.events.get(AnchorCreated);
    anchorCreated.add(context, {
      submitter: context.transaction.senderAddress,
      anchorID,
      createdAt,
    }, [context.transaction.senderAddress]);
  }
}
