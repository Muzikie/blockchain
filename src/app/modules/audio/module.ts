/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { TokenMethod, VerifyStatus } from 'lisk-framework';
import { CreateCommand } from './commands/create_command';
import { DestroyCommand } from './commands/destroy_command';
import { TransferCommand } from './commands/transfer_command';
import { StreamCommand } from './commands/stream_command';
import { ReclaimCommand } from './commands/reclaim_command';
import { SetAttributesCommand } from './commands/set_attributes_command';
import { AudioCreated } from './events/audioCreated';
import { AudioStreamed } from './events/audioStreamed';
import { AudioIncomeReclaimed } from './events/audioIncomeReclaimed';
import { AudioEndpoint } from './endpoint';
import {
  accountStoreSchema,
  audioStoreSchema,
  idRequestSchema,
  addressRequestSchema,
} from './schemas';
import { AudioMethod } from './method';
import { AudioAccountStore } from './stores/audioAccount';
import { AudioStore } from './stores/audio';
import { CollectionMethod } from '../collection/method';
import { SubscriptionMethod } from '../subscription/method';
import { COMMANDS, MODULES } from '../../constants';

export class AudioModule extends BaseModule {
public endpoint = new AudioEndpoint(this.stores, this.offchainStores);
public method = new AudioMethod(this.stores, this.events);

private readonly _createCommand = new CreateCommand(this.stores, this.events);
private readonly _destroyCommand = new DestroyCommand(this.stores, this.events);
private readonly _transferCommand = new TransferCommand(this.stores, this.events);
private readonly _setAttributesCommand = new SetAttributesCommand(this.stores, this.events);
private readonly _streamCommands = new StreamCommand(this.stores, this.events);
private readonly _reclaimCommands = new ReclaimCommand(this.stores, this.events);

// eslint-disable-next-line @typescript-eslint/member-ordering
public commands = [
  this._createCommand,
  this._destroyCommand,
  this._transferCommand,
  this._setAttributesCommand,
  this._streamCommands,
  this._reclaimCommands,
];

private _collectionMethod!: CollectionMethod;
private _subscriptionMethod!: SubscriptionMethod;
private _tokenMethod!: TokenMethod;

public constructor() {
  super();
  this.stores.register(AudioAccountStore, new AudioAccountStore(this.name, 0));
  this.stores.register(AudioStore, new AudioStore(this.name, 1));
  this.events.register(AudioCreated, new AudioCreated(this.name));
  this.events.register(AudioStreamed, new AudioStreamed(this.name));
  this.events.register(AudioIncomeReclaimed, new AudioIncomeReclaimed(this.name));
}

public addDependencies(
  collectionMethod: CollectionMethod,
  subscriptionMethod: SubscriptionMethod,
  tokenMethod: TokenMethod,
): void {
  this._collectionMethod = collectionMethod;
  this._subscriptionMethod = subscriptionMethod;
  this._tokenMethod = tokenMethod;

  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._streamCommands.addDependencies(this._subscriptionMethod);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._reclaimCommands.addDependencies(this._tokenMethod);
}

public metadata(): ModuleMetadata {
  return {
    ...this.baseMetadata(),
    endpoints: [
      {
        name: this.endpoint.getAccount.name,
        request: addressRequestSchema,
        response: accountStoreSchema,
      },
      {
        name: this.endpoint.getAudio.name,
        request: idRequestSchema,
        response: audioStoreSchema,
      },
    ],
    assets: [],
  };
}

// Lifecycle hooks
public async init(_args: ModuleInitArgs): Promise<void> {
  // initialize this module when starting a node
}

public async insertAssets(_context: InsertAssetContext) {
  // initialize block generation, add asset
}

public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
  // verify block
}

// Lifecycle hooks
// eslint-disable-next-line @typescript-eslint/require-await
public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
  // verify transaction will be called multiple times in the transaction pool
  return {
    status: VerifyStatus.OK,
  };
}

public async beforeCommandExecute(context: TransactionExecuteContext): Promise<void> {
  const audioAccountSubStore = this.stores.get(AudioAccountStore);
  const audioSubStore = this.stores.get(AudioStore);

  if (
    context.transaction.command === COMMANDS.DESTROY &&
context.transaction.module === MODULES.AUDIO
  ) {
    const account = await audioAccountSubStore.get(context, context.transaction.senderAddress);
    const audioID = account.audio.audios[account.audio.audios.length - 1];
    if (audioID) {
      const { collectionID } = await audioSubStore.get(context, audioID);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      await this._collectionMethod.removeAudio(
        context.getMethodContext(),
        audioID,
        collectionID,
        context.transaction.senderAddress,
      );
    }
  }
}

public async afterCommandExecute(context: TransactionExecuteContext): Promise<void> {
  const audioAccountSubStore = this.stores.get(AudioAccountStore);
  const audioSubStore = this.stores.get(AudioStore);

  if (
    context.transaction.command === COMMANDS.CREATE &&
    context.transaction.module === MODULES.AUDIO
  ) {
    const account = await audioAccountSubStore.get(context, context.transaction.senderAddress);
    const audioID = account.audio.audios[account.audio.audios.length - 1];
    if (audioID) {
      const { collectionID } = await audioSubStore.get(context, audioID);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      await this._collectionMethod.addAudio(
        context.getMethodContext(),
        audioID,
        collectionID,
        context.transaction.senderAddress,
      );
    }
  }
}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
