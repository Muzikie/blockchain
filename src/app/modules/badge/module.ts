/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */

import {
  BaseModule,
  ModuleInitArgs,
  InsertAssetContext,
  BlockVerifyContext,
  TransactionVerifyContext,
  VerificationResult,
  ModuleMetadata,
  TransactionExecuteContext,
  // GenesisBlockExecuteContext,
  // BlockExecuteContext,
  // BlockAfterExecuteContext,
} from 'lisk-sdk';
// eslint-disable-next-line import/no-extraneous-dependencies
import { TokenMethod, VerifyStatus } from 'lisk-framework';
import { CreateCommand } from './commands/create_command';
import { DestroyCommand } from './commands/destroy_command';
import { TransferCommand } from './commands/transfer_command';
import { StreamCommand } from './commands/stream_command';
import { ReclaimCommand } from './commands/reclaim_command';
import { SetAttributesCommand } from './commands/set_attributes_command';
import { BadgeCreated } from './events/badgeCreated';
import { BadgeStreamed } from './events/badgeStreamed';
import { BadgeIncomeReclaimed } from './events/badgeIncomeReclaimed';
import { BadgeEndpoint } from './endpoint';
import {
  accountStoreSchema,
  badgeStoreSchema,
  idRequestSchema,
  addressRequestSchema,
} from './schemas';
import { BadgeMethod } from './method';
import { BadgeAccountStore } from './stores/badgeAccount';
import { BadgeStore } from './stores/badge';
import { AnchorMethod } from '../anchor/method';
// import { COMMANDS, MODULES } from '../../constants';

export class BadgeModule extends BaseModule {
public endpoint = new BadgeEndpoint(this.stores, this.offchainStores);
public method = new BadgeMethod(this.stores, this.events);

private readonly _createCommand = new CreateCommand(this.stores, this.events);
private readonly _destroyCommand = new DestroyCommand(this.stores, this.events);
private readonly _transferCommand = new TransferCommand(this.stores, this.events);
private readonly _setAttributesCommand = new SetAttributesCommand(this.stores, this.events);
private readonly _streamCommands = new StreamCommand(this.stores, this.events);
private readonly _reclaimCommands = new ReclaimCommand(this.stores, this.events);

// eslint-disable-next-line @typescript-eslint/member-ordering
public commands = [
  this._createCommand,
  this._destroyCommand,
  this._transferCommand,
  this._setAttributesCommand,
  this._streamCommands,
  this._reclaimCommands,
];

private _anchorMethod!: AnchorMethod;
private _tokenMethod!: TokenMethod;

public constructor() {
  super();
  this.stores.register(BadgeAccountStore, new BadgeAccountStore(this.name, 0));
  this.stores.register(BadgeStore, new BadgeStore(this.name, 1));
  this.events.register(BadgeCreated, new BadgeCreated(this.name));
  this.events.register(BadgeStreamed, new BadgeStreamed(this.name));
  this.events.register(BadgeIncomeReclaimed, new BadgeIncomeReclaimed(this.name));
}

public addDependencies(
  anchorMethod: AnchorMethod,
  tokenMethod: TokenMethod,
): void {
  this._anchorMethod = anchorMethod;
  this._tokenMethod = tokenMethod;

  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._streamCommands.addDependencies(this._anchorMethod);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  this._reclaimCommands.addDependencies(this._tokenMethod);
}

public metadata(): ModuleMetadata {
  return {
    ...this.baseMetadata(),
    endpoints: [
      {
        name: this.endpoint.getAccount.name,
        request: addressRequestSchema,
        response: accountStoreSchema,
      },
      {
        name: this.endpoint.getBadge.name,
        request: idRequestSchema,
        response: badgeStoreSchema,
      },
    ],
    assets: [],
  };
}

// Lifecycle hooks
public async init(_args: ModuleInitArgs): Promise<void> {
  // initialize this module when starting a node
}

public async insertAssets(_context: InsertAssetContext) {
  // initialize block generation, add asset
}

public async verifyAssets(_context: BlockVerifyContext): Promise<void> {
  // verify block
}

// Lifecycle hooks
// eslint-disable-next-line @typescript-eslint/require-await
public async verifyTransaction(_context: TransactionVerifyContext): Promise<VerificationResult> {
  // verify transaction will be called multiple times in the transaction pool
  return {
    status: VerifyStatus.OK,
  };
}

public async beforeCommandExecute(_context: TransactionExecuteContext): Promise<void> {
  // const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
  // const badgeSubStore = this.stores.get(BadgeStore);

//   if (
//     context.transaction.command === COMMANDS.DESTROY &&
// context.transaction.module === MODULES.AUDIO
//   ) {
//     const account = await badgeAccountSubStore.get(context, context.transaction.senderAddress);
//     const badgeID = account.badge.badges[account.badge.badges.length - 1];
//     if (badgeID) {
//       const { collectionID } = await badgeSubStore.get(context, badgeID);
//       // eslint-disable-next-line @typescript-eslint/no-unsafe-call
//       await this._collectionMethod.removeBadge(
//         context.getMethodContext(),
//         badgeID,
//         collectionID,
//         context.transaction.senderAddress,
//       );
//     }
//   }
}

public async afterCommandExecute(_context: TransactionExecuteContext): Promise<void> {
  // const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
  // const badgeSubStore = this.stores.get(BadgeStore);

  // if (
  //   context.transaction.command === COMMANDS.CREATE &&
  //   context.transaction.module === MODULES.AUDIO
  // ) {
  //   const account = await badgeAccountSubStore.get(context, context.transaction.senderAddress);
  //   const badgeID = account.badge.badges[account.badge.badges.length - 1];
  //   if (badgeID) {
  //     const { collectionID } = await badgeSubStore.get(context, badgeID);
  //     // eslint-disable-next-line @typescript-eslint/no-unsafe-call
  //     await this._collectionMethod.addBadge(
  //       context.getMethodContext(),
  //       badgeID,
  //       collectionID,
  //       context.transaction.senderAddress,
  //     );
  //   }
  // }
}

  // public async initGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async finalizeGenesisState(_context: GenesisBlockExecuteContext): Promise<void> {}

  // public async beforeTransactionsExecute(_context: BlockExecuteContext): Promise<void> {}

  // public async afterTransactionsExecute(_context: BlockAfterExecuteContext): Promise<void> {}
}
