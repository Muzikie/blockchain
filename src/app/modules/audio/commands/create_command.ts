/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AudioStore } from '../stores/audio';
import { AudioAccountStore } from '../stores/audioAccount';
import { CreateCommandParams, Audio, AudioAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { VALID_GENRES, MIN_RELEASE_YEAR } from '../constants';
import { getEntityID, verifyHash } from '../../../utils';
import { AudioCreated } from '../events/audioCreated'

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const { params, transaction } = context;
    const thisYear = new Date().getFullYear();
    const numericYear = Number(params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          `Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`,
        ),
      };
    }
    if (params.genre.some(item => item > VALID_GENRES.length)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Genres should be selected from the list of valid genres'),
      };
    }
    if (
      params.owners.length === 0 ||
      params.owners.some(item => item.shares < 1) ||
      params.owners.reduce((acc, item) => acc + item.shares, 0) !== 100
    ) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          'Owners should have a total of 100 shares. Each owner should have a positive number of shares.',
        ),
      };
    }

    const isHashGenuine = verifyHash(params.audioSignature, params.audioHash, transaction.senderPublicKey);
    if (!isHashGenuine) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The hash signature is not generated by the sender'),
      };
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the audio file
    const audioID = getEntityID(context.transaction);

    const audioAccountSubStore = this.stores.get(AudioAccountStore);
    const audioSubStore = this.stores.get(AudioStore);

    // Check uniqueness of the NFT
    const audioExists = await audioSubStore.has(context, audioID);
    if (audioExists) {
      throw new Error('You have already created this audio.');
    }

    // Create the Audio object and save it on the blockchain
    const owners = params.owners.map(owner => ({
      address: owner.address,
      shares: owner.shares,
      income: BigInt(0),
    }));

    const audioObject: Audio = {
      ...params,
      owners,
      creatorAddress: transaction.senderAddress,
    };

    // Store the hash of the audio object in the sender account
    const accountExists = await audioAccountSubStore.has(context, transaction.senderAddress);
    if (accountExists) {
      const senderAccount: AudioAccount = await audioAccountSubStore.get(
        context,
        transaction.senderAddress,
      );
      senderAccount.audio.audios = [...senderAccount.audio.audios, audioID];
      await audioAccountSubStore.set(context, transaction.senderAddress, senderAccount);
    } else {
      await audioAccountSubStore.set(context, context.transaction.senderAddress, {
        audio: {
          audios: [audioID],
        },
      });
    }

    // @todo Here we should check if the Audio is already uploaded using steganography methods
    // Store the audio object in the blockchain
    await audioSubStore.set(context, audioID, audioObject);

    // Emit a "New collection" event
    const audioCreated = this.events.get(AudioCreated);
    audioCreated.add(context, {
      creatorAddress: context.transaction.senderAddress,
      audioID,
    }, [context.transaction.senderAddress]);
  }
}
