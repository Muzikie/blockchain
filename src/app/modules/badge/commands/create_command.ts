/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { BadgeStore } from '../stores/badge';
import { BadgeAccountStore } from '../stores/badgeAccount';
import { CreateCommandParams, Badge, BadgeAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { TREASURY_ADDRESS } from '../../../constants';
import { BadgeCreated } from '../events/badgeCreated';

const getBadgeID = (params: CreateCommandParams) =>
  Buffer.concat([
    Buffer.from(params.type, 'utf8'),
    Buffer.from(params.rank.toString(), 'utf8'),
    Buffer.from(params.awardDate, 'utf8')],
  );

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const { transaction, params } = context;

    const badgeSubStore = this.stores.get(BadgeStore);

    if (!transaction.senderAddress.equals(TREASURY_ADDRESS)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('You are not authorized to create a anchor.'),
      };
    }

    // Ensure that there's only one badge (of each type) per day
    const badgeID = getBadgeID(params);
    const badgeExists = await badgeSubStore.has(context, badgeID);
    if (badgeExists) {
      throw new Error('You have already created this badge.');
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the badge file
    const badgeID = getBadgeID(params);

    const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
    const badgeSubStore = this.stores.get(BadgeStore);

    // Create Badge NFT
    const badgeNFT: Badge = {
      ...params,
      badgeID,
      prize: BigInt(0), // @todo calculate prize for the rank
      claimed: false,
    };

    // Store the hash of the badge object in the sender account
    const accountExists = await badgeAccountSubStore.has(context, transaction.senderAddress);
    if (accountExists) {
      const senderAccount: BadgeAccount = await badgeAccountSubStore.get(
        context,
        transaction.senderAddress,
      );
      senderAccount.badges = [...senderAccount.badges, badgeID];
      await badgeAccountSubStore.set(context, transaction.senderAddress, senderAccount);
    } else {
      await badgeAccountSubStore.set(context, context.transaction.senderAddress, {
        badges: [badgeID],
      });
    }

    // Store the badge object in the blockchain
    await badgeSubStore.set(context, badgeID, badgeNFT);

    // Emit a "New collection" event
    const badgeCreated = this.events.get(BadgeCreated);
    badgeCreated.add(context, {
      badgeID,
      prize: BigInt(0),
    }, [context.transaction.senderAddress]);
  }
}
