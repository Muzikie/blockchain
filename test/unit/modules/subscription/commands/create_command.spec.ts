import { CreateCommand } from '../../../../../src/app/modules/subscription/commands/create_command';
import { SubscriptionModule } from '../../../../../src/app/modules/subscription/module';

describe('CreateCommand', () => {
  let command: CreateCommand;
  const module = new SubscriptionModule();

  beforeEach(() => {
    command = new CreateCommand(module.stores, module.events);
  });

  describe('constructor', () => {
    it('should have valid name', () => {
      expect(command.name).toEqual('create');
    });

    it('should have valid schema', () => {
      expect(command.schema).toMatchSnapshot();
    });
  });

  describe('verify', () => {
    describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
  });

  describe('execute', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
  });
});
