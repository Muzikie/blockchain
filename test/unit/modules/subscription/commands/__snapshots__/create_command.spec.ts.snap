// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCommand constructor should have valid schema 1`] = `
{
  "$id": "subscription/create",
  "properties": {
    "maxMembers": {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
    "price": {
      "dataType": "uint64",
      "fieldNumber": 2,
    },
  },
  "required": [
    "maxMembers",
    "price",
  ],
  "title": "CreateAsset transaction asset for subscription module",
  "type": "object",
}
`;
